### PR TITLE
Plfm 5604 manager

### PR DESCRIPTION
**Pull request by nickgros**

### CODE REVIEW

**Comment by nickgros** [comment]
https://github.com/Sage-Bionetworks/Synapse-Stack-Builder/pull/268

File: lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/CloudServiceMultipartUploadDAOProviderImpl.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	private GoogleCloudStorageMultipartUploadDAO googleCloudStorageMultipartUploadDAO;
+
+	@Override
+	public CloudServiceMultipartUploadDAO getCloudServiceMultipartUploadDao(UploadType uploadType) {
+		switch (uploadType) {
+			case S3:
+				return s3MultipartUploadDAO;
+			case GOOGLECLOUDSTORAGE:
+				return googleCloudStorageMultipartUploadDAO;
+			case SFTP:
```
Comment (john-hill) [inline lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/CloudServiceMultipartUploadDAOProviderImpl.java]: just have default.

File: lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAO.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package org.sagebionetworks.upload.multipart;
+
+/**
+ * Abstraction for Google Cloud multi-part upload.
+ *
+ */
+public interface GoogleCloudStorageMultipartUploadDAO extends CloudServiceMultipartUploadDAO {
```
Comment (john-hill) [inline lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAO.java]: drop this interface (and the s3 one as well)

File: lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/CloudServiceMultipartUploadDAO.java [hunk 3]
```diff
@@ -1,57 +1,46 @@
...
-	 * @param partMD5Hex
-	 */
-	public void addPart(AddPartRequest request);
-
-	/**
-	 * Delete an object for the given bucket and key.
-	 * @param bucket
-	 * @param key
-	 */
-	public void deleteObject(String bucket, String key);
```
Comment (nickgros) [inline lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/CloudServiceMultipartUploadDAO.java:48]: In all cases, delete object can occur atomically when addPart succeeds. The method does not need to be exposed (and hides a distinction between how the AWS and Google upload implementations differ)

File: lib/lib-googlecloud/pom.xml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	</parent>
+	<artifactId>lib-googlecloud</artifactId>
+	<name>lib-googlecloud</name>
+	<packaging>jar</packaging>
+
+	<properties>
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+	</properties>
+
+	<dependencies>
```
Comment (nickgros) [inline lib/lib-googlecloud/pom.xml:19]: Remove any unused dependencies

File: lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAOImpl.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+		}
+		if (!Hex.encodeHexString(Base64.decodeBase64(uploadedPart.getMd5())).equals(request.getPartMD5Hex())) {
+			throw new IllegalArgumentException("The provided MD5 does not match the MD5 of the uploaded part.  Please re-upload the part.");
+		}
+		// The part was uploaded successfully
+	}
+
+	// Parts that are stitched by Synapse and not uploaded by the client don't need to be validated
+	// (and cannot be validated, since we don't know the md5 before stitching)
+	@WriteTransaction
```
Comment (john-hill) [inline lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAOImpl.java]: transaction annotation is ignored for non-public methods.

File: lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAOImpl.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	void addPartWithoutValidatingMd5(String uploadId, String bucket, String key, Long lowerBound, Long upperBound, Long totalNumberOfParts) {
+		multipartUploadComposerDAO.addPartToUpload(uploadId, lowerBound, upperBound);
+		if (lowerBound != 1 || !upperBound.equals(totalNumberOfParts)) { // If this is false, we have the entire file and no longer need to merge parts.
+			attemptToMergePart(Long.valueOf(uploadId), bucket, key, lowerBound, upperBound, totalNumberOfParts);
```
Comment (john-hill) [inline lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAOImpl.java]: can move this logic to the attempt to merge part

File: lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAOImpl.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+		}
+	}
+
+	@MandatoryWriteTransaction
+	void attemptToMergePart(Long uploadId, String bucket, String key, Long lowerBound, Long upperBound, Long totalNumberOfParts) {
+		PartRange newStitchedPartRange = MultipartUploadUtils.getRangeOfPotentialStitchTargets(lowerBound, upperBound, totalNumberOfParts);
```
Comment (john-hill) [inline lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAOImpl.java]: fix infinite loop.

File: lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAOImpl.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+				}
+
+				// Create a new part
+				googleCloudStorageClient.composeObjects(bucket,
+						MultipartUploadUtils.createPartKeyFromRange(key, expectedStitchPartRange.getLowerBound(), expectedStitchPartRange.getUpperBound()),
+						partKeys);
+
+				// Delete the old parts from Google Cloud and the database.
+				for (String part : partKeys) {
+					googleCloudStorageClient.deleteObject(bucket, part);
```
Comment (john-hill) [inline lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAOImpl.java:98]: move delete to complete multi-part upload, else error thrown deleting would make this part unrecoverable.

**Review by john-hill — APPROVED** [review]

**Pull request merged**